File: 01_simple_google_sso/flutter/lib/services/auth_service.dart
import 'dart:async';

import 'package:authors_collection/config/google_sign_in_config.dart';
import 'package:authors_collection/models/auth_result.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:google_sign_in/google_sign_in.dart';

class AuthService {
  AuthService({FirebaseAuth? firebaseAuth, GoogleSignIn? googleSignIn})
    : _firebaseAuth = firebaseAuth ?? FirebaseAuth.instance,
      _googleSignIn = googleSignIn ?? GoogleSignIn.instance;

  final FirebaseAuth _firebaseAuth;
  final GoogleSignIn _googleSignIn;

  bool _googleInitialized = false;

  Stream<User?> get authStateChanges => _firebaseAuth.authStateChanges();

  User? get currentUser => _firebaseAuth.currentUser;

  Future<void> initialize() async {
    if (_googleInitialized) {
      return;
    }

    await _googleSignIn.initialize(
      serverClientId: GoogleSignInConfig.resolvedServerClientId,
    );
    _googleInitialized = true;
  }

  Future<AuthResult> signInWithGoogle() async {
    try {
      await initialize();

      final GoogleSignInAccount googleUser = await _googleSignIn.authenticate();
      final GoogleSignInAuthentication googleAuth = googleUser.authentication;
      final String? idToken = googleAuth.idToken;

      if (idToken == null || idToken.isEmpty) {
        return const AuthResult(
          success: false,
          message:
              'Google no devolvió un token válido. Revisa la configuración del proyecto.',
        );
      }

      final credential = GoogleAuthProvider.credential(idToken: idToken);
      await _firebaseAuth.signInWithCredential(credential);

      return const AuthResult(success: true);
    } on GoogleSignInException catch (error) {
      return AuthResult(success: false, message: _googleErrorMessage(error));
    } on FirebaseAuthException catch (error) {
      return AuthResult(success: false, message: _firebaseErrorMessage(error));
    } catch (_) {
      return const AuthResult(
        success: false,
        message: 'No se pudo iniciar sesión. Intenta nuevamente.',
      );
    }
  }

  Future<AuthResult> signOut() async {
    try {
      await Future.wait<void>([
        _firebaseAuth.signOut(),
        _googleSignIn.signOut(),
      ]);

      return const AuthResult(success: true);
    } on FirebaseAuthException catch (error) {
      return AuthResult(success: false, message: _firebaseErrorMessage(error));
    } catch (_) {
      return const AuthResult(
        success: false,
        message: 'No se pudo cerrar la sesión. Intenta nuevamente.',
      );
    }
  }

  String _firebaseErrorMessage(FirebaseAuthException error) {
    switch (error.code) {
      case 'account-exists-with-different-credential':
        return 'Ya existe una cuenta con ese email usando otro método de acceso.';
      case 'invalid-credential':
        return 'Las credenciales recibidas no son válidas.';
      case 'network-request-failed':
        return 'No hay conexión disponible. Revisa tu red e intenta nuevamente.';
      case 'user-disabled':
        return 'Esta cuenta fue deshabilitada.';
      default:
        return error.message ??
            'Ocurrió un error autenticando contra Firebase.';
    }
  }

  String _googleErrorMessage(GoogleSignInException error) {
    switch (error.code) {
      case GoogleSignInExceptionCode.canceled:
        return 'Inicio de sesión cancelado.';
      case GoogleSignInExceptionCode.clientConfigurationError:
        return 'Google Sign-In no está configurado correctamente. Revisa SHA-1, google-services.json e Info.plist.';
      case GoogleSignInExceptionCode.providerConfigurationError:
        return 'El proveedor de Google no está disponible o está mal configurado.';
      case GoogleSignInExceptionCode.uiUnavailable:
        return 'No se pudo mostrar la interfaz de Google Sign-In.';
      case GoogleSignInExceptionCode.interrupted:
        return 'La autenticación fue interrumpida. Intenta nuevamente.';
      default:
        return error.description ??
            'Ocurrió un error al iniciar sesión con Google.';
    }
  }
}

File: 01_simple_google_sso/swift/google-sso-sample/AuthService.swift
import FirebaseAuth
import FirebaseCore
import GoogleSignIn
import UIKit

enum StartupConfigurationError: LocalizedError {
    case missingGoogleServiceInfo
    case invalidGoogleServiceInfo
    case placeholderGoogleServiceInfo
    case placeholderInfoPlist

    var errorDescription: String? {
        switch self {
        case .missingGoogleServiceInfo:
            return "Falta el archivo GoogleService-Info.plist en google-sso-sample/Resources."
        case .invalidGoogleServiceInfo:
            return "GoogleService-Info.plist existe, pero no tiene un formato valido."
        case .placeholderGoogleServiceInfo:
            return "GoogleService-Info.plist todavia tiene placeholders. Reemplazalo con el archivo real de Firebase."
        case .placeholderInfoPlist:
            return "Info.plist todavia tiene placeholders de Google Sign-In. Completa GIDClientID, GIDServerClientID y CFBundleURLSchemes."
        }
    }
}

final class AuthService {
    private var authStateHandle: AuthStateDidChangeListenerHandle?

    private var auth: Auth {
        Auth.auth()
    }

    deinit {
        if let authStateHandle {
            auth.removeStateDidChangeListener(authStateHandle)
        }
    }

    func configureFirebaseIfNeeded() throws {
        if FirebaseApp.app() != nil {
            return
        }

        guard let filePath = Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist") else {
            throw StartupConfigurationError.missingGoogleServiceInfo
        }

        guard let options = FirebaseOptions(contentsOfFile: filePath) else {
            throw StartupConfigurationError.invalidGoogleServiceInfo
        }
        
        guard let clientID = options.clientID,
              let apiKey = options.apiKey else {
            throw StartupConfigurationError.invalidGoogleServiceInfo
        }
        
        let hasPlaceholders = [
            options.googleAppID,
            clientID,
            apiKey,
            options.projectID ?? "",
        ].contains { value in
            value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            value.contains("YOUR_") ||
            value.contains("REPLACE_")
        }

        if hasPlaceholders {
            throw StartupConfigurationError.placeholderGoogleServiceInfo
        }

        let infoClientID = Bundle.main.object(forInfoDictionaryKey: "GIDClientID") as? String ?? ""
        let infoServerClientID = Bundle.main.object(forInfoDictionaryKey: "GIDServerClientID") as? String ?? ""
        let urlTypes = Bundle.main.object(forInfoDictionaryKey: "CFBundleURLTypes") as? [[String: Any]] ?? []
        let urlSchemes = urlTypes.flatMap { item in
            item["CFBundleURLSchemes"] as? [String] ?? []
        }

        let hasInfoPlistPlaceholders = [infoClientID, infoServerClientID].contains { value in
            value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || value.contains("YOUR_")
        } || urlSchemes.contains(where: { $0.contains("YOUR_") })

        if hasInfoPlistPlaceholders {
            throw StartupConfigurationError.placeholderInfoPlist
        }

        FirebaseApp.configure(options: options)
    }

    func observeAuthState(_ onChange: @escaping (AuthUser?) -> Void) {
        if let authStateHandle {
            auth.removeStateDidChangeListener(authStateHandle)
        }

        authStateHandle = auth.addStateDidChangeListener { _, user in
            onChange(user.map(Self.makeAuthUser(from:)))
        }
    }

    @MainActor
    func signInWithGoogle() async -> Result<Void, Error> {
        do {
            try configureFirebaseIfNeeded()

            guard let rootViewController = UIApplication.shared.topViewController else {
                throw NSError(
                    domain: "google-sso-sample.auth",
                    code: 1001,
                    userInfo: [NSLocalizedDescriptionKey: "No se encontro una pantalla valida para presentar Google Sign-In."]
                )
            }

            guard let clientID = FirebaseApp.app()?.options.clientID, !clientID.isEmpty else {
                throw NSError(
                    domain: "google-sso-sample.auth",
                    code: 1002,
                    userInfo: [NSLocalizedDescriptionKey: "Falta el Client ID de iOS en la configuracion de Firebase."]
                )
            }

            let serverClientID = Bundle.main.object(forInfoDictionaryKey: "GIDServerClientID") as? String
            //let normalizedServerClientID = serverClientID?.trimmingCharacters(in: .whitespacesAndNewlines)
            GIDSignIn.sharedInstance.configuration = GIDConfiguration(
                clientID: clientID,
                //serverClientID: normalizedServerClientID?.isEmpty == false ? normalizedServerClientID : nil
            )

            let result = try await GIDSignIn.sharedInstance.signIn(withPresenting: rootViewController)
            let googleUser = result.user

            guard let idToken = googleUser.idToken?.tokenString else {
                throw NSError(
                    domain: "google-sso-sample.auth",
                    code: 1003,
                    userInfo: [NSLocalizedDescriptionKey: "Google no devolvio un token valido. Revisa la configuracion del proyecto."]
                )
            }

            let credential = GoogleAuthProvider.credential(
                withIDToken: idToken,
                accessToken: googleUser.accessToken.tokenString
            )

            _ = try await auth.signIn(with: credential)
            return .success(())
        } catch {
            return .failure(mapToUserReadableError(error))
        }
    }

    func signOut() -> Result<Void, Error> {
        do {
            try auth.signOut()
            GIDSignIn.sharedInstance.signOut()
            return .success(())
        } catch {
            return .failure(error)
        }
    }

    private func mapToUserReadableError(_ error: Error) -> Error {
        let nsError = error as NSError

        if nsError.localizedDescription.localizedCaseInsensitiveContains("canceled") ||
            nsError.localizedDescription.localizedCaseInsensitiveContains("cancelled") {
            return NSError(
                domain: nsError.domain,
                code: nsError.code,
                userInfo: [NSLocalizedDescriptionKey: "Inicio de sesion cancelado."]
            )
        }

        if nsError.domain == AuthErrorDomain {
            switch AuthErrorCode(_bridgedNSError: nsError)?.code {
            case .accountExistsWithDifferentCredential:
                return NSError(
                    domain: nsError.domain,
                    code: nsError.code,
                    userInfo: [NSLocalizedDescriptionKey: "Ya existe una cuenta con ese email usando otro metodo de acceso."]
                )
            case .invalidCredential:
                return NSError(
                    domain: nsError.domain,
                    code: nsError.code,
                    userInfo: [NSLocalizedDescriptionKey: "Las credenciales recibidas no son validas."]
                )
            case .networkError:
                return NSError(
                    domain: nsError.domain,
                    code: nsError.code,
                    userInfo: [NSLocalizedDescriptionKey: "No hay conexion disponible. Revisa tu red e intenta nuevamente."]
                )
            case .userDisabled:
                return NSError(
                    domain: nsError.domain,
                    code: nsError.code,
                    userInfo: [NSLocalizedDescriptionKey: "Esta cuenta fue deshabilitada."]
                )
            default:
                break
            }
        }

        return NSError(
            domain: nsError.domain,
            code: nsError.code,
            userInfo: [NSLocalizedDescriptionKey: nsError.localizedDescription]
        )
    }

    private static func makeAuthUser(from user: FirebaseAuth.User) -> AuthUser {
        AuthUser(
            displayName: user.displayName?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
                ? user.displayName!
                : "Usuario sin nombre",
            email: user.email?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
                ? user.email!
                : "Sin email disponible",
            photoURL: user.photoURL?.absoluteString
        )
    }
}

private extension UIApplication {
    var topViewController: UIViewController? {
        let rootViewController = connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first(where: \.isKeyWindow)?
            .rootViewController

        return topViewController(from: rootViewController)
    }

    func topViewController(from viewController: UIViewController?) -> UIViewController? {
        if let navigationController = viewController as? UINavigationController {
            return topViewController(from: navigationController.visibleViewController)
        }

        if let tabBarController = viewController as? UITabBarController {
            return topViewController(from: tabBarController.selectedViewController)
        }

        if let presentedViewController = viewController?.presentedViewController {
            return topViewController(from: presentedViewController)
        }

        return viewController
    }
}
